From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sun, 18 Aug 2024 18:07:24 +0800
Subject: [PATCH] Disable respawn ceiling

This feature is not yet complete, but no major bugs have been found.

Note that you should set spawnRadius to at least 1, not 0, and the spawn point range should be at least 3*3 (or more)

diff --git a/src/main/java/net/minecraft/server/level/PlayerRespawnLogic.java b/src/main/java/net/minecraft/server/level/PlayerRespawnLogic.java
index e1ee330b6fdd43114a1330677b1597bf437c1ed5..4c0cd2432a99b0d0b8f80eb46bff17b24ff6ede1 100644
--- a/src/main/java/net/minecraft/server/level/PlayerRespawnLogic.java
+++ b/src/main/java/net/minecraft/server/level/PlayerRespawnLogic.java
@@ -7,6 +7,7 @@ import net.minecraft.core.Direction;
 import net.minecraft.core.SectionPos;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.levelgen.Heightmap;
@@ -17,14 +18,29 @@ public class PlayerRespawnLogic {
     protected static BlockPos getOverworldRespawnPos(ServerLevel world, int x, int z) {
         boolean bl = world.dimensionType().hasCeiling();
         LevelChunk levelChunk = world.getChunk(SectionPos.blockToSectionCoord(x), SectionPos.blockToSectionCoord(z));
-        int i = bl ? world.getChunkSource().getGenerator().getSpawnHeight(world) : levelChunk.getHeight(Heightmap.Types.MOTION_BLOCKING, x & 15, z & 15);
-        if (i < world.getMinBuildHeight()) {
-            return null;
+        // Vine start - disableRespawnCeiling
+        int i ;
+        if (bl) {
+            i = world.getChunkSource().getGenerator().getSpawnHeight(world);
         } else {
-            int j = levelChunk.getHeight(Heightmap.Types.WORLD_SURFACE, x & 15, z & 15);
-            if (j <= i && j > levelChunk.getHeight(Heightmap.Types.OCEAN_FLOOR, x & 15, z & 15)) {
-                return null;
+            // Vine start - Safe respawn
+            if (one.tranic.vine.config.VineConfig.Feature.disableRespawnCeiling) {
+                i = world.levelData.getSpawnPos().getY();
+                for (int w = 0; w < 10; w++) {
+                    BlockPos block = new BlockPos(x, i, z);
+                    if (world.getBlockState(block).getBlock() == Blocks.AIR) {
+                        break;
+                    }
+                    i = i + 3;
+                }
+                // Vine end - Safe respawn
             } else {
+                i = levelChunk.getHeight(Heightmap.Types.MOTION_BLOCKING, x & 15, z & 15);
+            }
+        }
+        if (i >= world.getMinBuildHeight()) {
+            int j = levelChunk.getHeight(Heightmap.Types.WORLD_SURFACE, x & 15, z & 15);
+            if (j > i || j <= levelChunk.getHeight(Heightmap.Types.OCEAN_FLOOR, x & 15, z & 15)) {
                 BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos();
 
                 for (int k = i + 1; k >= world.getMinBuildHeight(); k--) {
@@ -42,6 +58,8 @@ public class PlayerRespawnLogic {
                 return null;
             }
         }
+        return null;
+        // Vine end - disableRespawnCeiling
     }
 
     @Nullable
@@ -50,16 +68,16 @@ public class PlayerRespawnLogic {
             return null;
         } else {
             return world.chunkScheduler.getRegionLocker().lockRegion(RegionPos.forChunk(chunkPos), () -> { // ShreddedPaper - lock region
-            for (int i = chunkPos.getMinBlockX(); i <= chunkPos.getMaxBlockX(); i++) {
-                for (int j = chunkPos.getMinBlockZ(); j <= chunkPos.getMaxBlockZ(); j++) {
-                    BlockPos blockPos = getOverworldRespawnPos(world, i, j);
-                    if (blockPos != null) {
-                        return blockPos;
+                for (int i = chunkPos.getMinBlockX(); i <= chunkPos.getMaxBlockX(); i++) {
+                    for (int j = chunkPos.getMinBlockZ(); j <= chunkPos.getMaxBlockZ(); j++) {
+                        BlockPos blockPos = getOverworldRespawnPos(world, i, j);
+                        if (blockPos != null) {
+                            return blockPos;
+                        }
                     }
                 }
-            }
 
-            return null;
+                return null;
             }); // ShreddedPaper
         }
     }
